man/imagemap.1
.Dd March 26, 1996
.Dt IMAGEMAP 1
.Os xs-httpd/3.5
.Sh NAME
.Nm imagemap
.Nd The WWW image mapper
.Sh SYNOPSIS
.Nm http://your.site.name/cgi\-bin/imagemap/~ Ns Ar username Ns / Ns Pa mapfile
.Sh DESCRIPTION
.Ql imagemap
is a program that lets you handle imagemaps on the web. An
imagemap is basically a picture (for example a GIF file)
that has been divided into areas. A user with a browser that
understands imagemaps (most modern graphical browser do)
can then click on a certain part of the image. An
area\-specific action then occurs.
.Pp
For example, consider a GIF picture with two areas on it: a
box on the left side and a circle on the right side. Then
the user who created the pictures can define what will
happen if the remote user clicks on the box, what will
happen if the circle is pressed, and also what happens if
the user clicks on the picture, but not on the box or the
circle (the default).
.Pp
.Xr imagemap 1
is not a program that allows the user to define the map.
There are other programs that do this; for example,
.Xr mapedit 1
is a good one.
.Xr imagemap
is the program that handles the actions for the webserver.
.Pp
To make sure that the
.Xr imagemap 1
program will have access to the
.Ar mapfile
in a user's private webdirectory, it should be included in the
.Sy PathInfoScripts
option of the
.Xr httpd.conf 5
configuration file, with its relative pathname
.Pa /cgi\-bin/imagemap/ .
This lets the script be executed with the privileges of the
user whose directory is accessed.
.Sh EXAMPLES
What a user must do is create a map file (either by hand,
which is not very easy, or by using a program like mapedit).
Then, he or she must use something like the following in
his/her webpage:
.Pp
.Bd -literal -offset indent -compact
<A\ HREF="http://site/cgi\-bin/imagemap/~username/pics/image.map">
 <IMG SRC="pics/image.gif" ISMAP>
</A>
.Ed
.Pp
In other words, make a
.Sy href
anchor to the imagemapper and use an image as the contents
of the anchor. Do not forgot to specify the
.Sy ismap
keyword in the
.Sy img
tag.
.Sh SEE ALSO
.Xr httpd 1 ,
.Xr httpd_cgi 7
.Pp
The project homepage:
.Pa http://www.xs\-httpd.org/
.Sh HISTORY
This program is derived from the NCSA httpd distribution,
which can be found at
.Pa http://hoohoo.ncsa.uiuc.edu/cgi/ .
It was cleaned up and converted for inclusing in the
xs\-httpd distribution in 1995.


ChangeLog
This file contains an overview of version differences. The list
is not complete. Refer to cvs comments for more information.

Changes since 3.7 beta 8
- New ServerInfo option to set version string (full, branch, name)
- Pass along all HTTP headers when a forwarding proxy request

Changes since 3.7 beta 7
- Rename hier.layout to dirs.conf and shuffle configurable dirs

Changes since 3.7 beta 6
- Fix IconDir/ExecDir usage for vhosts
- Fix handling of unprivileged uid
- Fix mod_python for versions other than 2.5
- Fix mod_gzip dependency
- Fix mod_ruby initialization (enable component loading)
- Easier install directory configuration via hier.layout
- Abort configure when a selected package cannot be used

Changes since 3.7 beta 5
- Execute in .xsconf now works if script.methods is emtpy
- UseLocalScript now defaults to on (and is checked properly)
- Fix on-the-fly (de)compression: now works with large files

Changes since 3.7 beta 4
- Add UseGzipCompression to mod_gzip for on-the-fly compression
- Build mod_bzip2 by default

Changes since 3.7 beta 3
- Fix bug that caused internal: commands in .xsscripts to be ignored
- Add mod_gzip, mod_compress, mod_bzip to replace external binaries
- Add Content-MD5 trailers for CGI scripts

Changes since 3.7 beta 2
- Fix authentication module logic: all modules need to grant access

Changes since 3.7 beta 1
- Modules are now implemented as shared libraries
- Add Modules option to set which modules get loaded

Changes since 3.6 beta 24
- Fix LDAP authentication
- Introduce seperated modules (mod_ldap, perl, python, ruby)
- Make OpenSSL mandatory (removes 40 #ifdefs, 2 libcompat files)

Changes since 3.6 beta 23
- Clean up 'forward' syntax in .redir to match 'redir' command
- Include a replacement crypt() if libc, libcrypt and libssl fail
- Add NoPrivs option to .xsconf (and deprecate .xsuid)
- Enable Smashing Stack Protector (SSP) by default
- Fix parsing of empty lines in mime.types

Changes since 3.6 beta 22
- Fix possible build conflict caused by libmd and openssl
- Fix parsing of mime.types files

Changes since 3.6 beta 21
- Fix memory mismanagement in digest authentication verification
- Replace fgets(), fgetln() by friendlier fparseln() routine
  Include replacement functions for fgetln() and fparseln()
- Fix crash in https timeout handler
- Portability fixes (b20 doesn't work on many platforms)
- Tweak TCP options (especially for Linux systems)
- Add xspasswd -r option to remove accounts from password file

Changes since 3.6 beta 20
- Remove arbitrary string length limits for many configuration files,
  HTTP header messages and SSL certificate fields
- Add support for SSLVhosts (using TLS extension: Server Name Indication)
- Fix NULL-pointer dereference and string termination in HTML encoding
- Add UseSendfile option in httpd.conf to disable sendfile(2)

Changes since 3.6 beta 19
- Fix memory mismanagement in digest authentication verification

Changes since 3.6 beta 18
- Enable experimental ruby interpreter (internal:ruby)

Changes since 3.6 beta 17
- Fix parsing problem in run-cgi SSI (and maybe other places)

Changes since 3.6 beta 16
- Add index.xhtml and index.xml to default IndexFiles
- Improve url-encoding routine

Changes since 3.6 beta 15
- Fix escaping of user input used in HTML error messages;
  possible XSS abuse reported by Serge van den Boom
- Fix checks for pcre used by libxs

Changes since 3.6 beta 14
- Fix potential crash in server-side include parsing

Changes since 3.6 beta 13
- Always terminate SSL connection when reading from socket fails
- Add conditional ifenv option to .redir files

Changes since 3.6 beta 12
- Don't install empty docdir
- Shuffle internal structures (reduce global vars, getenv calls)
- Reject non-HTTP protocol specifiers

Changes since 3.6 beta 11
- Drop persistent connections after client read/write errors

Changes since 3.6 beta 10
- Add RedirFile configuration option for light-weight vhosts
- Sort qvalues in accept-* arrays
- Fake qvalues for */* and type/* when qvalues are not used

Changes since 3.6 beta 9
- Use original URL in rewriting rules before trying index.html
- Add PhIconDir option, analogous to PhExecDir

Changes since 3.6 beta 8
- Close unused CGI pipes (could block execution), by jilles

Changes since 3.6 beta 7
- Extend characters available for alternative setproctitle()
- Lock pidfile and don't start if somebody holds a lock
- Allow multiple arguments to Restrict and AuthFile
- Allow method selector in .xsconf blocks (e.g. only match POST)

Changes since 3.6 beta 6
- Improve configuration checks
- Fix include order and typechecks

Changes since 3.6 beta 5
- Enable FastCGI support again (experimental)
- Clean Perl and Python detection and initialisation (experimental)

Changes since 3.6 beta 4
- Shuffle configuration files
- Fix build outside source directory (wrt config.h)
- Fix sendfile on Linux
- Fix infinate read when content-length exceeds post data
- Portability fixes in function/library checking

Changes since 3.6 beta 3
- Remove outputbuffer and set TCP_NOPUSH (where available)
- Add 'passexist' redirection command

Changes since 3.6 beta 2
- Fix server-side include variable cleanup
- Use sendfile (where appropriate)

Changes since 3.6 beta 1
- Add http status code back to $ERROR_READABLE

Changes since 3.5 beta 31
- No functional changes

Changes since 3.5 beta 30
- Fix read errors with long POST content

Changes since 3.5 beta 29
- Make CGI header size dynamic (unlimited)
- Add default PathInfoScripts (imagemap, xschpass)
- Update manpages

Changes since 3.5 beta 28
- Fix basic authentication check
- Fix ldap authentication check

Changes since 3.5 beta 27
- Fix reading of stderr data from CGI scripts (could block CGI)
- Minor code cleanups

Changes since 3.5 beta 26
- Add -f filename option to xspasswd
- Change default of UsePut to true; document default settings
- Minor portability improvements
- Minor changes in server error messages (standard format)

src/imagemap.c
/*
	imagemap.c - taken from the NCSA httpd distribution, which can
	be found at http://hoohoo.ncsa.uiuc.edu/cgi/

	Previous authors:
	- Kevin Hughes (kevinh@pulua.hcc.hawaii.edu)
	- Eric Haines (erich@eye.com)
	- Rob McCool (robm@ncsa.uiuc.edu)
	- Chris Hyams (cgh@rice.edu)
	- Craig Milo Rogers (rogers@isi.edu)
	- Carlos Varela (cvarela@ncsa.uiuc.edu)

	Changes: cleaned up the code, used the XS-HTTPD defines,
	removed old code which is no longer needed.

	This version by Sven Berkvens (sven@stack.nl).
*/

#include	"config.h"

#include	<stdio.h>
#include	<sys/types.h>
#include	<stdlib.h>
#include	<ctype.h>
#include	<sys/stat.h>
#include	<string.h>

#include	"decode.h"
#include	"constants.h"

#define		MAXVERTS	1000
#define		X		0
#define		Y		1

static	int	isname			(int) WARNUNUSED;
static	void	servererr		(const char *)	NORETURN;
static	int	pointinpoly		(void) WARNUNUSED;
static	int	pointincircle		(void) WARNUNUSED;
static	int	pointinrect		(void) WARNUNUSED;
static	void	sendmesg		(const char *)	NORETURN;

static	double	testpoint[2], pointarray[MAXVERTS][2];

int
main(int argc, char **argv)
{
	char		input[MYBUFSIZ], *mapname, def[MYBUFSIZ],
			errstr[MYBUFSIZ], *t, mapname2[MYBUFSIZ];
	const	char	*query;
	int		i, j, k, sawpoint = 0;
	FILE		*fp;
	double		dist, mindist = 0;

	if (!(query = getenv("QUERY_STRING")))
		servererr("Invalid usage, client may not support ISMAP");

	if (!(t = strchr(query, ',')))
		servererr("Your client does not support image mapping");
	*t++ = 0;
	i = atoi(query);
	testpoint[X] = (double)i;
	i = atoi(t);
	testpoint[Y] = (double)i;

	if (!(mapname = getenv("PATH_TRANSLATED")))
		servererr("No translated path given by server");
	if (!(fp = fopen(mapname, "r")))
	{
		snprintf(mapname2, MYBUFSIZ, "/vwww/httpd/%s", mapname);
		mapname = mapname2;
		if (!(fp = fopen(mapname, "r")))
		{
			snprintf(errstr, MYBUFSIZ,
				"Could not open map file: `%s'", mapname);
			servererr(errstr);
		}
	}

	while (fgets(input, MYBUFSIZ, fp))
	{
		char		type[MYBUFSIZ], url[MYBUFSIZ], num[10];
		size_t		length;

		length = strlen(input);
		while ((length > 0) && (input[length - 1] <= ' '))
			input[--length] = 0;
		if ((input[0] == '#') || (!input[0]))
			continue;

		type[0] = url[0] = 0;

		for (i = 0; isname(input[i]) && (input[i]); i++)
			type[i] = input[i];
		type[i] = 0;

		while (isspace(input[i]))
			i++;
		for (j = 0; input[i] && isname(input[i]); ++i, ++j)
			url[j] = input[i];
		url[j] = 0;

		if (!strcmp(type, "default") && !sawpoint)
		{
			strlcpy(def, url, MYBUFSIZ);
			continue;
		}

		k = 0;
		while (input[i])
		{
			while (isspace(input[i]) || (input[i] == ','))
				i++;
			j = 0;
			while (isdigit(input[i]))
				num[j++] = input[i++];
			num[j] = '\0';
			if (num[0])
			{
				j = atoi(num);
				pointarray[k][X] = (double)j;
			} else
				break;
			while (isspace(input[i]) || (input[i] == ','))
				i++;
			j = 0;
			while (isdigit(input[i]))
				num[j++] = input[i++];
			num[j] = 0;
			if (num[0])
			{
				j = atoi(num);
				pointarray[k++][Y] = (double)j;
			} else
				servererr("Missing y value.");
			if (k == (MAXVERTS - 1))
				break;
		}
		pointarray[k][X] = -1;
		if (!strcasecmp(type, "poly"))
		{
			if (pointinpoly())
				sendmesg(url);
		}
		if (!strcasecmp(type, "circle"))
		{
			if (pointincircle())
				sendmesg(url);
		}
		if (!strcasecmp(type, "rect"))
		{
			if (pointinrect())
				sendmesg(url);
		}
		if (!strcasecmp(type, "point"))
		{
			dist = ((testpoint[X] - pointarray[0][X]) *
				(testpoint[X] - pointarray[0][X])) +
				((testpoint[Y] - pointarray[0][Y]) *
				(testpoint[Y] - pointarray[0][Y]));
			if ((!sawpoint) || (dist < mindist))
			{
				mindist = dist;
				strlcpy(def, url, MYBUFSIZ);
			}
			sawpoint++;
		}
	}
	if (def[0])
		sendmesg(def);
	servererr("No default specified");
	/* NOTREACHED */
	(void)argc;
	(void)argv;
}

static	void
sendmesg(const char *url)
{
	printf("Location: %s\nContent-type: text/html\n\n", url);
	printf("<HTML><HEAD><TITLE>Moved</TITLE></HEAD><BODY><H1>Moved</H1>\n");
	printf("This document has <A HREF=\"%s\">moved</A>\n", url);
	printf("</BODY></HTML>\n");
	exit(1);
}

static	int
pointinrect()
{
	return ((testpoint[X] >= pointarray[0][X]) &&
		(testpoint[X] <= pointarray[1][X]) &&
		(testpoint[Y] >= pointarray[0][Y]) &&
		(testpoint[Y] <= pointarray[1][Y]));
}

static	int
pointincircle()
{
	int		radius1, radius2;

	radius1 =	((pointarray[0][Y] - pointarray[1][Y]) *
			 (pointarray[0][Y] - pointarray[1][Y])) +
			((pointarray[0][X] - pointarray[1][X]) *
			 (pointarray[0][X] - pointarray[1][X]));
	radius2 =	((pointarray[0][Y] - testpoint[Y]) *
			 (pointarray[0][Y] - testpoint[Y])) +
			((pointarray[0][X] - testpoint[X]) *
			 (pointarray[0][X] - testpoint[X]));
	return (radius2 <= radius1);
}

static	int
pointinpoly()
{
	int		i, numverts, xflag0, crossings;
	double		*p, *stop, tx, ty, y;

	for (i = 0; (pointarray[i][X] != -1) && (i < MAXVERTS); i++)
		/* NOTHING HERE */;
	numverts = i; crossings = 0;
	tx = testpoint[X]; ty = testpoint[Y];
	y = pointarray[numverts - 1][Y];

	p = (double *)pointarray + 1;
	if ((y >= ty) != (*p >= ty))
	{
		if ((xflag0 = (pointarray[numverts - 1][X] >= tx)) ==
			(*(double *)pointarray >= tx))
		{
			if (xflag0)
				crossings++;
		} else
		{
			crossings += ((pointarray[numverts - 1][X] - (y - ty) *
				(*(double *)pointarray - pointarray[numverts - 1][X]) /
				(*p - y)) >= tx);
		}
	}

	stop = pointarray[numverts];

	for (y = *p, p += 2; p < stop; y = *p, p += 2)
	{
		if (y >= ty)
		{
			while ((p < stop) && (*p >= ty))
				p += 2;
			if (p >= stop)
				break;
			if ((xflag0 = (*(p - 3) >= tx)) == (*(p - 1) >= tx))
			{
				if (xflag0)
					crossings++;
			} else
			{
				crossings += ((*(p - 3) - (*(p - 2) - ty) *
					(*(p - 1) - *(p - 3)) /
					(*p - *(p - 2))) >= tx);
			}
		} else
		{
			while ((p < stop) && (*p < ty))
				p += 2;
			if (p >= stop)
				break;
			if ((xflag0 = (*(p - 3) >= tx)) == (*(p - 1) >= tx))
			{
				if (xflag0)
					crossings++;
			} else
			{
				crossings += ((*(p - 3) - (*(p - 2) - ty) *
					(*(p - 1) - *(p - 3)) /
					(*p - *(p - 2))) >= tx);
			}
		}
	}
	return(crossings & 1);
}

static	void
servererr(const char *msg)
{
	printf("Content-type: text/html\n\n");
	printf("<HTML><HEAD><TITLE>Mapping server error</TITLE></HEAD>\n");
	printf("<BODY><H1>Mapping server error</H1>\n");
	printf("The mapping server encountered an error:<P>\n");
	printf("%s", escape(msg));
	printf("</BODY></HTML>\n");
	exit(1);
}

static	int
isname(int c)
{
	return(!isspace(c));
}

